IndexedDB Tile Cache

andremig-bentley opened the pull request:
Create a local cache with IndexedDB which will be used to store tiles from the mesh export service. The IndexedDBCache class is a standalone class which interfaces with IndexedDB and can be used widely. The class can store, delete, and retrieve content from an IndexedDB database, and has the option to have the content in the database expire. 

If the useIndexedDBCache conditional is set true, tiles being fetched through the itwinjs-batched-models channel will first attempt to fetch their content from an IndexedDB database. If that content cannot be found or is expired, the channel will fetch the content normally and store the content for future requests.

andremig-bentley commented on core/frontend/src/tile/IDBTileStorage.ts:
```diff
@@ -0,0 +1,182 @@
+
+import { TileRequestChannel } from "./internal";
+
+export class IDBStorage {
```
IDBTileStorage has been changed to IDBStorage, a more generic class that has no mention of tiles. All of its functionality deals solely with the IndexedDB database itself.

andremig-bentley commented on core/frontend/src/tile/IDBTileStorage.ts:
```diff
@@ -0,0 +1,182 @@
+    return this._tileStorage;
+  }
+
+  public async fetch(url: string, callback: (url: string) => Promise<Response>): Promise<ArrayBuffer> {
```
This fetch function now takes a callback function so that it can always return content even if the db request is unsuccessful. For now, I've kept this function as a part of the tile request channel, but it can be moved.

danieliborra replied:
This couples things more. Try removing completely IDBTileRequestChannel

andremig-bentley replied:
I decided to keep this fetch function as simple as possible to read since it is the one public function, meaning that instead of having the actual cache searching in this function, I kept the cache searching separate. I think that if the fetch and searchCache functions were combined it would be a much more complicated function to read, but if you think they should be combined, I can certainly do that.

andremig-bentley commented on extensions/frontend-tiles/src/BatchedTile.ts:
```diff
@@ -137,8 +140,9 @@ export class BatchedTile extends Tile {
     url.search = this.batchedTree.reader.baseUrl.search;
-    const response = await fetch(url.toString());
-    return response.arrayBuffer();
+    const response = await this.channel.fetch(url.toString(), fetch);
```
Here is the usage of the new fetch function. For now, I kept this as a part of the tile request channel. No matter what, to access the db like this, we need these tiles to have a member with access to the db. Since the channel already existed and has access to the db, I kept this functionality as a part of the channel. It might be better to revert the channel being used by the batched tiles to a standard TileRequestChannel, and instead give the tiles a member that is an instance of IDBStorage, or some other connector class that communicates between the two. This shouldn't be a difficult change, but needs further discussion.

danieliborra replied:
As we discussed, I'd keep this simple, no need to add additional channels here, there is already a channel.
I'd remove the new channel. 
Branch this branch and let's try it out without channels, you will see the code will be more simple and elegant IMO.

andremig-bentley commented:
@danieliborra 

Updates based on your feedback. I believe the IDBStorage class should now be testable on its own, so I'm working on getting that going.

danieliborra commented on core/frontend/src/tile/IDBTileStorage.ts:
```diff
@@ -0,0 +1,182 @@
+    };
+  }
+
+  public async addContent(uniqueId: string, content: ArrayBuffer) {
```
make it private

danieliborra commented on core/frontend/src/tile/IDBTileStorage.ts:
```diff
@@ -0,0 +1,182 @@
+    return undefined;
+  }
+
+  public async deleteContent(uniqueId: string) {
```
make it private

danieliborra commented on core/frontend/src/tile/IDBTileStorage.ts:
```diff
@@ -0,0 +1,182 @@
+    await this._db.close();
+  }
+
+  public async requestContent(uniqueId: string): Promise<ArrayBuffer | undefined> {
```
call it "fetch"

danieliborra commented on extensions/frontend-tiles/src/BatchedTile.ts:
```diff
@@ -137,8 +140,9 @@ export class BatchedTile extends Tile {
     url.search = this.batchedTree.reader.baseUrl.search;
-    const response = await fetch(url.toString());
-    return response.arrayBuffer();
+    const response = await this.channel.fetch(url.toString(), fetch);
```
```suggestion
    const response = await this.cache.fetch(url.toString(), fetch);
```

danieliborra commented on core/frontend/src/tile/IDBTileStorage.ts:
```diff
@@ -0,0 +1,182 @@
+/** @packageDocumentation
+ * @module Tiles
+ */
+
```
Can you create an interface e.g. `IStorageCache` with a virtual function `fetch` and make IDBStorage inherit from it?

Then, as step too, can you create another class `IDBPassThrowCache`, which just recalls fetch? We will need both classes in the future.

danieliborra reviewed (changes requested):
Good progress, but we're still not there. As I commented below, I'd suggest branching this branch and see how it looks when removing the new channel.
It won't take you long, and then we can discuss. I think it is worth a try.

andremig-bentley commented:
IDBTileRequestChannel has now been fully removed, IDBStorage has been renamed to IndexedDBCacheManager (copilot's suggestion). The interface IndexedDBCache has been created with a fetch function, and IndexedDBCacheManager and IndexedDBCachePassThrough implement the interface. Within IndexedDBCacheManager, all functions except for the fetch function have been made private. Within BatchedTiles.ts, the IDBTileRequestChannel has been replaced with _localCache, an instance of IndexedDBCacheManager.

danieliborra commented:
> IDBTileRequestChannel has now been fully removed, IDBStorage has been renamed to IndexedDBCacheManager (copilot's suggestion). The interface IndexedDBCache has been created with a fetch function, and IndexedDBCacheManager and IndexedDBCachePassThrough implement the interface. Within IndexedDBCacheManager, all functions except for the fetch function have been made private. Within BatchedTiles.ts, the IDBTileRequestChannel has been replaced with _localCache, an instance of IndexedDBCacheManager.

It looks way better! All changes are good.
In this case, co-pilot didn't do great job 😄 
What about:
- IndexedDBCache => ILocalCache or ICloudContentCache or IFetchContentCache (generic name, has nothing to do with IndexDB)
- IndexedDBCacheManager => IndexedDBCache
- IndexedDBCachePassThrough => PassThroughCache

danieliborra commented on extensions/frontend-tiles/src/BatchedTile.ts:
```diff
@@ -29,6 +29,7 @@ export class BatchedTile extends Tile {
   private readonly _unskippable: boolean;
   /** Transform from the tile's local coordinate system to that of the tileset. */
   public readonly transformToRoot?: Transform;
+  private _localCache: any;
```
```suggestion
  private _localCache: ILocalCache;
```

andremig-bentley commented:
Updates from most recent commit:

- Fixed all async issues/issues with database functions happening out of order, etc. Confirmed now that all database functionality is working as intended (searching, retrieving, adding, deleting, all working how/when they're supposed to).
- General cleanup, changed IDBTileStorage.ts to IndexedDBCache.ts
- Moved all changes dealing with IModelTileRequestChannels to a new branch and removed them from this PR

markschlosseratbentley reviewed (changes requested):
extract-api

andremig-bentley commented:
> extract-api

Taken care of.

andremig-bentley commented on core/frontend/src/tile/TileAdmin.ts:
```diff
@@ -161,6 +161,8 @@ export class TileAdmin {
   public readonly contextPreloadParentSkip: number;
   /** @beta */
   public readonly cesiumIonKey?: string;
+  /** @beta */
```
Potential location for new conditional to use/not use cache, can definitely be moved if somewhere else is better.

markschlosseratbentley commented:
is extract-api failing again?

andremig-bentley commented:
> is extract-api failing again?

Yeah it is, not sure why. I re-ran it and re-committed everything applicable with my latest commit, but it doesn’t seem to be working. Looking into that this morning

markschlosseratbentley commented:
> > is extract-api failing again?
> 
> Yeah it is, not sure why. I re-ran it and re-committed everything applicable with my latest commit, but it doesn’t seem to be working. Looking into that this morning

All good.  Approving

pmconne commented on common/api/core-frontend.api.md:
```diff
@@ -5989,6 +5989,12 @@ export class IdleTool extends InteractiveTool {
 }
 
+// @beta (undocumented)
+export interface ILocalCache {
```
> Do not use I as a prefix for interface names.

Please make sure you have reviewed our [coding guidelines](https://www.itwinjs.org/learning/guidelines/typescript-coding-guidelines/#names)

pmconne commented on common/api/core-frontend.api.md:
```diff
@@ -7365,6 +7371,23 @@ export class IndexBuffer {
     toVertexIndices(): VertexIndices;
 }
 
+// @beta (undocumented)
```
beta APIs must be documented.

pmconne commented on core/frontend/src/test/tile/IndexedDBCache.test.ts:
```diff
@@ -0,0 +1,126 @@
+  }
+}
+
+describe.only("IndexedDBCache", () => {
```
rm .only

pmconne reviewed (changes requested):
If this is no longer "WIP" then please update your PR description to summarize what you are doing and why.

pmconne commented on core/frontend/src/tile/IndexedDBCache.ts:
```diff
@@ -0,0 +1,185 @@
+    }
+  }
+
+  protected async open(dbCache: IndexedDBCache){
```
This is an instance method, why does it take an IndexedDBCache argument instead of operating on `this`?

pmconne commented on core/frontend/src/tile/IndexedDBCache.ts:
```diff
@@ -0,0 +1,185 @@
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+
+/** @beta */
```
Why do we want these APIs to be part of the package's public API? They seem like implementation details.

pmconne commented on core/frontend/src/tile/IndexedDBCache.ts:
```diff
@@ -0,0 +1,185 @@
+      };
+
+      storedResponse.onerror = async () => {
+        await this.close();
```
Log errors.

andremig-bentley commented on core/frontend/src/test/tile/IndexedDBCache.test.ts:
```diff
@@ -0,0 +1,126 @@
+    const content = await cache.convertStringToBuffer(uniqueId);
+
+    await expiringCache.testAddContent(uniqueId, content);
+    await pause(1000);
```
Do you mean because of the pause?

andremig-bentley replied:
I changed the test to now have an expiration time of -1 so that it expires immediately with no reliance on a pause. I changed the test where the cache does not expire to have a much larger expiration time to avoid flakiness in that test as well.

pmconne commented on core/frontend/src/tile/TileAdmin.ts:
```diff
@@ -1274,6 +1277,13 @@ export namespace TileAdmin { // eslint-disable-line no-redeclare
+     * Default value: false
+     * @alpha
+     */
+    useIndexedDBCache?: boolean;
```
Can you move this to `FrontendTilesOptions`?

andremig-bentley replied:
Moved it to `FrontendTilesOptions` and added it wherever necessary within the rest of the options functionality. Let me know if I missed something or incorrectly configured some of it.

pmconne commented on extensions/frontend-tiles/src/FrontendTiles.ts:
```diff
@@ -25,6 +25,9 @@ function createMeshExportServiceQueryUrl(args: { iModelId: string, urlPrefix?: s
     url = `${url}&cdn=1`;
 
+  if (args.useIndexedDBCache)
+    url = `${url}&idb=1`;
```
Why do you need to alter the URL based on whether or not the cache is used?

pmconne replied:
Now you can remove `useIndexedDBCache` from the two `Args` types.

andremig-bentley replied:
Yeah I saw that too, just pushed that change a minute ago

pmconne commented on extensions/frontend-tiles/src/BatchedTile.ts:
```diff
@@ -29,6 +30,7 @@ export class BatchedTile extends Tile {
   private readonly _unskippable: boolean;
   /** Transform from the tile's local coordinate system to that of the tileset. */
   public readonly transformToRoot?: Transform;
+  private _localCache: LocalCache;
```
```suggestion
  private readonly _localCache: LocalCache;
```

pmconne commented:
Force-merging because unrelated failing reality data integration test.
